Add support for social login with Facebook, Google

timothyej opened the pull request:
Fixes #44.

This PR adds services and directives to facilitate the implementation of social login with Stormpath.
### Notes
- Only support for Facebook and Google initially as Github and LinkedIn has different flows
- Google Redirect URI must be set to `postmessage`
### Dependencies
- https://github.com/stormpath/express-stormpath/pull/121
### How it works
- The `SocialLogin` service asks the back-end (i.e. express-stormpath) for a list of available OAuth providers.
- It then maps them to each service, i.e. `Facebook` to the `facebookLogin` service.
- The `spLoginForm` directive will list all enabled providers.
- A click on e.g. the `Login with Facebook` button will trigger the popup.
- Once the user has authorised, they will be logged in using the `/login` route in express-stormpath.
### Includes
- `socialLogin` service that provides the `getProviders` method to get a list of OAuth providers and their Angular services
- `facebookLogin` service for login with Facebook
- `googleLogin` service for login with Google
- `spSocialLogin` directive that adds a button attached to a social provider, e.g.
  
  ``` html
  <button sp-social-login="facebook" sp-scope="public_profile,email">Login with Facebook</button>
  ```
- Edits to the login form to include social login buttons
### How to verify
1. Checkout this branch and build it using `grunt build --force`
2. Copy the files in the `dist/` folder to your own Angular app with Stormpath integrated
3. Make sure that you have an Express API up and running with the [`social-login`](https://github.com/stormpath/express-stormpath/pull/121) branch
4. Create and configure Stormpath Directories for both Facebook and Google
5. Verify that login buttons for Facebook and Google show up on the login form (`spLoginForm`)
6. Verify that it works to login with both providers, and also that the username/password flow still works
7. Verify that logout works as expected
### Screenshots

**Login:**  
<img width="586" alt="screen shot 2015-10-16 at 04 20 41" src="https://cloud.githubusercontent.com/assets/721091/10541139/91cf3bcc-73c3-11e5-9d0e-9230731362b1.png">

**Registration:**  
<img width="543" alt="screen shot 2015-10-21 at 21 14 30" src="https://cloud.githubusercontent.com/assets/721091/10647629/bb57e21c-7839-11e5-8294-97e594184267.png">

timothyej committed:
add support for social login with facebook, google, and linkedin

robertjd commented on src/stormpath.social-login.js:
```diff
@@ -0,0 +1,143 @@
+ *
+ * <pre>
+ * <div class="container">
+ *   <button sp-social-login="facebook" sp-scope="public_profile,email">Login with Facebook</button>
```
This is awesome, +1

timothyej replied:
Thank you :relieved:

robertjd commented on src/stormpath.login.js:
```diff
@@ -2,7 +2,12 @@
-.controller('SpLoginFormCtrl', ['$scope','$auth',function ($scope,$auth) {
+.controller('SpLoginFormCtrl', ['$scope','$auth','$socialLogin',function ($scope,$auth,$socialLogin) {
+  // Load list of social login providers from server.
+  $socialLogin.getProviders().then(function(providers) {
```
Because of the network request, the provider buttons won't appear immediately (I think this is a bad experience).  I think that the `getProviders` function should cache the HTTP result.  If I want that resource to be pre-fetched (before the login form is rendered) I could put a call to `getProviders` in a `app.run()` block

timothyej replied:
Yes, that's a good point! :+1:

rorheden commented on src/stormpath.social-login.facebook.js:
```diff
@@ -0,0 +1,88 @@
+    var clientId;
+
+    function FacebookLoginService() {
+      return this;
```
This is not needed. Also, a "constructor" should never return anything except itself (default).

rorheden commented on src/stormpath.social-login.linkedin.js:
```diff
@@ -0,0 +1,66 @@
+    var clientId;
+    var auth2;
+
+    function LinkedinLoginService() {
```
This should be declared outside of the provider.

rorheden commented on src/stormpath.social-login.linkedin.js:
```diff
@@ -0,0 +1,66 @@
+
+angular.module('stormpath.linkedinLogin', [])
+.provider('$linkedinLogin', function() {
+  function loadLinkedinSDK(clientId, onload) {
```
`Linkedin` should be `LinkedIn`and `onload` should be `onLoad`.

rorheden commented on src/stormpath.social-login.linkedin.js:
```diff
@@ -0,0 +1,66 @@
+  function loadLinkedinSDK(clientId, onload) {
+    var id = 'linkedin-jssdk';
+    var firstJs = document.getElementsByTagName('script')[0];
+    var js;
```
`js` is very unclear. I'd name it something more clear like `jsElement`.

rorheden commented on src/stormpath.social-login.js:
```diff
@@ -0,0 +1,143 @@
+
+      $socialLogin.getProviders().then(function(providers) {
+        provider = providers[attrs.spSocialLogin].service;
+        provider && provider.init(element);
```
I prefer clarity/readability instead of one-liners. I.e. I'd rather have this be:

```
if (provider) {
  provider.init(element);
}
```

rorheden commented on src/stormpath.social-login.js:
```diff
@@ -0,0 +1,143 @@
+      var provider;
+
+      $socialLogin.getProviders().then(function(providers) {
+        provider = providers[attrs.spSocialLogin].service;
```
This is actually not a provider, but a provider service... Right? Also, no checking if `attrs.spSocialLogin` actually exists in `providers`.

rorheden commented on src/stormpath.social-login.js:
```diff
@@ -0,0 +1,143 @@
+    link: function(scope, element, attrs) {
+      var provider;
+
+      $socialLogin.getProviders().then(function(providers) {
```
There is no handing if the promise would fail.

rorheden commented on src/stormpath.social-login.js:
```diff
@@ -0,0 +1,143 @@
+          } else if (err.data && err.data.error) {
+            scope.error = err.data.error;
+          } else {
+            scope.error = 'XHR Error';
```
This could be more descriptive.

rorheden commented on src/stormpath.social-login.linkedin.js:
```diff
@@ -0,0 +1,66 @@
+    js.id = id;
+    js.src = '//platform.linkedin.com/in.js';
+    js.onload = onload;
+    js.innerHTML = 'api_key: ' + clientId + '\n' +
```
Very unclear what this is... This should be broken out to a separate method and linked to LinkedIn's tag format. I.e. something like, createLinkedInScriptTag(clientId, callback).

rorheden commented on src/stormpath.social-login.linkedin.js:
```diff
@@ -0,0 +1,66 @@
+    js.src = '//platform.linkedin.com/in.js';
+    js.onload = onload;
+    js.innerHTML = 'api_key: ' + clientId + '\n' +
+                   'onLoad: ' + onload.name + '\n' +
```
This does not work for anonymous methods. Would be nicer to wrap the callback and generate a function that you attach on the window scope, then use the name of that function.

rorheden commented on src/stormpath.social-login.linkedin.js:
```diff
@@ -0,0 +1,66 @@
+    LinkedinLoginService.prototype.name = 'LinkedIn';
+
+    LinkedinLoginService.prototype.clientId = function(value) {
+      if (value) {
```
No validation on exactly what is a valid value. Is `true` a valid clientId?

rorheden commented on src/stormpath.social-login.linkedin.js:
```diff
@@ -0,0 +1,66 @@
+      IN.User.authorize(function() {
+        deferred.resolve({
+          providerData: {
+              providerId: 'linkedin',
```
Inconsistent indentation. Should be 2 spaces and not 4.

rorheden commented on src/stormpath.social-login.js:
```diff
@@ -0,0 +1,143 @@
+          try {
+            service = $injector.get('$' + providerName + 'Login');
+          } catch (err) {
+            return;
```
What happens if there is an error here?

timothyej replied:
This is if there has been addad a new social provider in Stormpath that hasn't been added to the SDK yet.

rorheden replied:
Ok. Though I think that `getProviders()` is doing a bit too much. What it actually does is not only _get providers_, it _initializes_ them. So I'd rather have it as two separate methods. `initializeProviders()` that sets up all available providers and produces a list of initialized providers, and `getProviders()` that reads from that previous list.

rorheden commented on src/stormpath.social-login.js:
```diff
@@ -0,0 +1,143 @@
+          }
+
+          service.clientId(provider.clientId);
+          providers[providerName].service = service;
```
`provider` is declared as `providers[providerName]` but `providers[providerName]` is still used here?

rorheden commented on src/stormpath.social-login.google.js:
```diff
@@ -0,0 +1,73 @@
+      auth2.grantOfflineAccess(options).then(function(response) {
+        deferred.resolve({
+          providerData: {
+              providerId: 'google',
```
Inconsistent indentation, 4 spaces instead of 2.

rorheden commented on src/stormpath.social-login.js:
```diff
@@ -0,0 +1,143 @@
+        scope.posting = true;
+
+        provider.login(options).then(function(data) {
+          return $auth.authenticate(data);
```
`return` is not needed here?

timothyej replied:
Yes it is, it returns a promise. If this promise fails, it will trigger the `catch` function to catch the error.

rorheden replied:
I stand corrected:) :beers:

rorheden commented on src/stormpath.social-login.linkedin.js:
```diff
@@ -0,0 +1,66 @@
+    var js;
+
+    if (document.getElementById(id)) {
+      return;
```
This should probably be handled, by either calling `onload` with an error or throwing an exception.

robertjd commented:
FYI, I just created a spec document in our framework specification.  Please follow this PR in case we change any details about how the configuration object or default URLs are used:

https://github.com/stormpath/stormpath-framework-spec/pull/9

rorheden commented on src/stormpath.social-login.linkedin.js:
```diff
@@ -0,0 +1,66 @@
+
+    LinkedinLoginService.prototype.name = 'LinkedIn';
+
+    LinkedinLoginService.prototype.clientId = function(value) {
```
Inconsistent. All other functions are named, but this is anonymous.

robertjd commented on src/stormpath.login.js:
```diff
@@ -14,7 +21,7 @@ angular.module('stormpath')
       .catch(function(response){
         $scope.posting = false;
-        $scope.error = response.data && response.data.error || 'XHR Error';
+        $scope.error = response.data && response.data.error || 'An error occured when communicating with server.';
```
For a future PR, let's create an HTTP transform method which applies this default error message.

robertjd commented on src/stormpath.social-login.google.js:
```diff
@@ -0,0 +1,84 @@
+
+      GoogleLoginService.prototype.name = 'Google';
+
+      GoogleLoginService.prototype.clientId = function clientId(value) {
```
I think this method should be named `setClientId`.  It also feels not-DRY to be repeating this function in every provider service.

rorheden commented on src/stormpath.social-login.facebook.js:
```diff
@@ -0,0 +1,67 @@
+    }
+  }
+
+  function FacebookLoginService($window, $q, $spJsLoader) {
```
Instead of declaring it as one function that returns an object, I would add the init and login methods as prototypes of FacebookLoginService. I.e:

``` js
function FacebookLoginService($window, $q, $spJsLoader) {
  this.$window = $window;
  this.$q = $q;
  this.$spJsLoader = $spJsLoader;
  this.name = 'Facebook';
  this.clientId = null;
}

FacebookLoginService.prototype.init = function init() {
  // ...
};

FacebookLoginService.prototype.login = function login(options) {
  // ...
};
```

Also, `new FacebookLoginService()` wouldn't give you an instance of `FacebookLoginService`.

timothyej replied:
Very good point, will fix :+1:

rorheden commented on src/stormpath.social-login.js:
```diff
@@ -0,0 +1,187 @@
+
+        if (document.getElementById(id)) {
+          deferred.resolve();
+          return deferred.promise;
```
I'd remove the `return deferred.promise;` and instead wrap the trailing lines 122-127 in an `else`.

rorheden commented on src/stormpath.social-login.linkedin.js:
```diff
@@ -0,0 +1,43 @@
+  'use strict';
+
+  function loadLinkedInSDK($spJsLoader, clientId) {
+    var innerHTML = 'api_key: ' + clientId + '\n' +
```
This doesn't say anything about the format. I'd also have it as a separate builder method.

``` js
// Comment describing where to reference the LinkedIn config format.
function buildLinkedInJsConfig(clientId) {
  return 'api_key: ' + clientId + '\nauthorize: true';
}
```

rorheden commented on src/stormpath.social-login.js:
```diff
@@ -0,0 +1,187 @@
+      });
+    }
+
+    return {
```
Same as mentioned before. Bad practice to have a constructor return another object than itself. Since `new SocialLoginService() != instanceof SocialLoginService`.

rorheden commented on src/stormpath.social-login.facebook.js:
```diff
@@ -0,0 +1,67 @@
+(function() {
+  'use strict';
```
Strict is only applied to the scoped part, and not the whole file. Not better to have it first?

timothyej replied:
Since files are concatenated and minified in production, it's best to specify the strict mode for this scope only: http://yuiblog.com/blog/2010/12/14/strict-mode-is-coming-to-town/

rorheden replied:
Ah, that is a really good answer! I did not think of that:) +1

robertjd replied:
+1

rorheden commented on src/stormpath.social-login.google.js:
```diff
@@ -0,0 +1,61 @@
+    this.$spJsLoader = $spJsLoader;
+  }
+
+  angular.extend(GoogleLoginService.prototype, {
```
Feels very unclear why `angular.extend` is used. Move the properties to the constructor and the functions as explicit prototype declarations.

rorheden commented on src/stormpath.social-login.facebook.js:
```diff
@@ -0,0 +1,67 @@
+  FacebookLoginService.prototype.login = function login(options) {
+    var deferred = this.$q.defer();
+
+    FB.login(loginCallback.bind(this, deferred), options);
```
Any reason why you're binding it with this?

timothyej replied:
No, this can actually be bind to `null`, which I will do :+1:

rorheden replied:
Great, makes it a little clearer. :+1:

rorheden commented on src/module.js:
```diff
@@ -83,7 +83,15 @@
  * </pre>
  */
-angular.module('stormpath',['stormpath.CONFIG','stormpath.auth','stormpath.userService'])
+angular.module('stormpath',[
```
Inconsistent. Missing space between `',` and `['`. Other module declarations have a space. Or if not having a space is the "consistent" way to go, then the other module declarations are the ones that are inconsistent.

timothyej replied:
In my files I follow our conventions, but here I just added new dependencies to the list. However, will add the space.

robertjd commented:
My initial review of this PR is positive, though I do want to see it working end-to-end before we continue.  I'm also getting the 7200 error from our when I try to post the codes from Google.  Let's dig into this after our upcoming standup

timothyej commented:
I got Google to work by setting the `redirect_uri` to `postmessage` in the Stormpath Directory.

However, LinkedIn's flow [is a bit different](https://developer-programs.linkedin.com/documents/exchange-jsapi-tokens-rest-api-oauth-tokens) from the other providers, so I decided to disable support for LinkedIn while we look into how to solve that case.

Instead, I will look into adding support for GitHub.

timothyej commented:
GitHub doesn't offer any good SDKs to do the login flow like Facebook and Google does. Instead I would have to create my own code for this or just use their redirect flow.

I will skip this and only support Facebook and Google in the first version and evaluate how this goes. Then later I will add support for GitHub and LinkedIn by probably using their redirect flows instead. It's inconsistent but it's most likely the best way to solve this.

Will work on the design and some logout problems, then this should be ready to test, and eventually merge.

timothyej committed:
remove support for linkedin due to big differences in their oauth flow

https://developer-programs.linkedin.com/documents/exchange-jsapi-tokens-rest-api-oauth-tokens

timothyej committed:
redesign the social login buttons

rorheden commented on src/stormpath.auth.js:
```diff
@@ -135,6 +135,7 @@ angular.module('stormpath.auth',['stormpath.CONFIG'])
         op.then(function(){
           $rootScope.$broadcast(STORMPATH_CONFIG.SESSION_END_EVENT);
         },function(response){
+          $rootScope.$broadcast(STORMPATH_CONFIG.SESSION_END_EVENT);
```
What type of error would cause this (i.e. connection, internal server error?)? Isn't it better to handle that specific error? Uncertain if really any error should end the session... (actually, I'm not even certain it ends the session, or how the event is handled... just guessing:))

timothyej replied:
One thing that might cause this is for example when the user's session has timed out. When they try to logout, the server returns an error that the token is invalid and the user can not logout from the UI.

timothyej replied:
I removed this again as I solved the error in the express-stormpath SDK.

rorheden commented:
Made a little comment, otherwise I think it looks good :) :+1:

robertjd commented on src/stormpath.social-login.js:
```diff
@@ -0,0 +1,197 @@
+          }
+        });
+
+        element.click(function() {
```
`element.click` appears to require the proper jQuery library, I think if we use `element.bind('click')` we can avoid that (using angular's jqLite)

timothyej replied:
Good catch, will update!

timothyej committed:
set accept header to application/json when doing logout

timothyej committed:
use element.bind instead of element.click to be compatible with jqLite in AngularJS

timothyej commented:
This is ready for final review. Please follow my instructions in the description and let me know if you need any boilerplate to get started.

robertjd commented on src/stormpath.social-login.js:
```diff
@@ -0,0 +1,197 @@
+    this.providersPromise = providersPromise;
+
+    this.$http.get(this.STORMPATH_CONFIG.getUrl('OAUTH_PROVIDERS_ENDPOINT')).then(function(response) {
+      var providers = response.data;
```
If the developer hasn't upgraded to the dependent version of Express, `providers` will be an HTML response and this will cause an exception on line 34 in this file.  We don't want to break someone's application, so there should be a check here to assert that this is the expected JSON structure.  If not, fail silently by returning an empty providers array.  The changelog also needs to be updated to inform the user that they must upgrade to the latest version of the Express library if they wish to use the social login feature.

timothyej replied:
Done :+1: Haven't edited the changelog yet though.

robertjd commented on src/stormpath.config.js:
```diff
@@ -169,6 +186,24 @@ angular.module('stormpath.CONFIG',[])
+    * Used by {@link stormpath.socialLoginService.$socialLogin#getProviders $socialLogin.getProviders()}.
+    *
+    */
+    OAUTH_PROVIDERS_ENDPOINT: '/oauth/providers',
```
We want to change this URL, as the returned data structure doesn't actually have anything to do with OAuth authentication.  My suggestion is to rename this endpoint to `/spa-config` and have an object in it with a property of `socialProviders`.  Thoughts?

timothyej replied:
How about `/spa-config/social-providers` so that we can keep the code as it is without generalising too much?

robertjd commented on src/stormpath.social-login.js:
```diff
@@ -0,0 +1,197 @@
+   *
+   * @description
+   *
+   * Add this directive to a button or link in order to authenticate using an OAuth provider. The value should be an OAuth provider id such as google or facebook.
```
Ping @omgitstom - I want to make sure we're using the correct language to refer to the social providers.  Should we be calling it an OAuth provider, social provider, or directory provider?

robertjd replied:
Per our chat during standup today, we want to call these "social providers"

timothyej replied:
Done :+1:

robertjd commented on src/stormpath.social-login.js:
```diff
@@ -0,0 +1,197 @@
+   *
+   * @name stormpath.spSocialLogin:spSocialLogin
+   *
+   * @description
```
@timothyej We should add a note to this section about the need for `postmessage`.  I'd just copy in this note from our REST product guide:

> **NOTE:** If you are using Google+ Sign-In for server-side apps, Google recommends that you leave the Authorized redirect URI field blank in the Google Developer Console. In Stormpath, when creating the Google Directory, you must set the redirect URI to postmessage.

And also add a link to this documentation for more information: http://docs.stormpath.com/guides/social-integrations/

timothyej replied:
:+1:

robertjd commented:
@timothyej this is looking great! :+1:  Please see new comments today

timothyej committed:
rename oauth providers to social providers and change endpoint to /spa-config/social-provider

timothyej commented:
@robertjd thanks for the feedback! I've addressed your comments in my last commit, and regarding the end-point, what about `/spa-config/social-providers`? If we add more configs later, we could merge them into one endpoint (e.g. `/spa-config`).

robertjd commented:
We do have more config to add later, for example the password policy for the default account store (so that you can do client-side validation of password strength).  So I think the endpoint should be generic, as the response will eventually look like:

```
{
  socialProviders: {},
  passwordPolicy: {}
}
```

timothyej committed:
change endpoint to /spa-config

timothyej commented:
@robertjd Got it! I've updated this PR and the express PR with these changes :)

timothyej committed:
add social login to registration form

timothyej commented:
I've also added the social login buttons to the registration form. The CSS used to style the buttons is duplicated for each form, and there's also some code in the controllers that is a bit redundant. I will address these issues in separate PRs later.
